### PR TITLE
View only available associations for product

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                      |
| Deprecations?   | no  |
| License         | MIT                                                          |

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 1]
```diff
@@ -230,4 +232,38 @@ public function findByTaxon(TaxonInterface $taxon): array
...
+            ->setParameter('code', $code)
+            ->setParameter('enabled', true)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+
+        if (null === $product) {
+            $product = $this->findOneByChannelAndCode($channel, $code);
+            $product->getAssociations()->clear();
+        }
```
Comment (TheMilek) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php:255]: Similar case to https://github.com/Sylius/Sylius/pull/14643#discussion_r1048562328

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 2]
```diff
@@ -232,4 +232,39 @@ public function findByTaxon(TaxonInterface $taxon): array
...
+            ->setParameter('enabled', true)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+
+        if (null === $product) {
+            $product = $this->findOneByChannelAndCode($channel, $code);
+
+            $product?->getAssociations()->clear();
+        }
```
Comment (TheMilek) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php:256]: Similar case to https://github.com/Sylius/Sylius/pull/14643#discussion_r1048562328

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 3]
```diff
@@ -232,4 +232,39 @@ public function findByTaxon(TaxonInterface $taxon): array
...
+            ->setParameter('code', $code)
+            ->setParameter('enabled', true)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+
+        if (null === $product) {
+            $product = $this->findOneByChannelAndCode($channel, $code);
+
+            $product?->getAssociations()->clear();
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php:255]:
```suggestion
            $product = $this->findOneByChannelAndCode($channel, $code);
            if (null === $product) {
                return null;
            }

            $product->getAssociations()->clear();
```
No point in even trying to apply any additional logic if the product cannot be not found at all.

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/AvailableProductAssociationsInProductCollectionExtensionSpec.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        $queryBuilder->addSelect('association')->shouldBeCalled()->willReturn($queryBuilder);
+        $queryBuilder->leftJoin('o.associations', 'association')->shouldBeCalled()->willReturn($queryBuilder);
+        $expr->andX(Argument::type(Expr\Comparison::class), Argument::type(Expr\Comparison::class))->willReturn($andx);
+        $expr->eq('associatedProduct.enabled', 'true')->shouldBeCalled()->willReturn($comparison);
+        $expr->eq('association.owner', 'o')->shouldBeCalled()->willReturn($comparison);
+        $queryBuilder->expr()->willReturn($expr->getWrappedObject());
+        $queryBuilder->leftJoin('association.associatedProducts', 'associatedProduct', 'WITH', Argument::type(Andx::class))->shouldBeCalled()->willReturn($queryBuilder);
+        $queryBuilder->andWhere('o.associations IS EMPTY OR associatedProduct.id IS NOT NULL')->shouldBeCalled()->willReturn($queryBuilder);
+        $queryBuilder->andWhere('o.enabled = :enabled')->shouldBeCalled()->willReturn($queryBuilder);
+        $queryBuilder->setParameter('enabled', true)->shouldBeCalled()->willReturn($queryBuilder);
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/AvailableProductAssociationsInProductCollectionExtensionSpec.php:94]: Space it out a bit

**Comment by GSadee** [comment]
Thank you, Kamil! :tada:

**Pull request merged**